### PR TITLE
adding restricted dates property for datetime calendar

### LINKED ISSUE #7598: Calendar: Cherry pick restricted dates to date-time package calendar
#7509

### PR DESCRIPTION
**Pull request by lorejoh12**
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #7598 
- [x] Include a change request file using `$ npm run change`

#### Description of changes

Cherry-picking the changes for restricted dates to the date-time package calendar

#### Focus areas to test

(optional)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/8089)

### CODE REVIEW

**Comment by JasonGore** [comment]
I think it would be good to add both snapshot tests and unit tests since this PR affects render output and date selection functionality.

**Comment by lorejoh12** [comment]
I also found a bug with restricted dates being on the borders of the grid, and it breaks arrowing completely. Working on fixing it right now.

**Comment by lorejoh12** [comment]
arrowing bug is fixed

**Comment by lorejoh12** [comment]
> I think it would be good to add both snapshot tests and unit tests since this PR affects render output and date selection functionality.

I totally agree, but this isn't the office ui fabric calendar, this is porting functionality to the date-time calendar (which isn't published yet).

we have existing work items to make all the snapshots and the unit tests, #6045 and #8094 (I couldn't find an existing item for this so I made one, but it's on our radar). Both of these are out of scope for this change. :)

File: packages/date-time/src/components/Calendar/examples/Calendar.Inline.Example.tsx [hunk 1]
```diff
@@ -38,6 +38,7 @@ export interface ICalendarInlineExampleProps {
   showWeekNumbers?: boolean;
```
Comment (cliffkoh) [inline packages/date-time/src/components/Calendar/examples/Calendar.Inline.Example.tsx]: line 5/6 in this file should probably use the full package import (examples should avoid using relative import)

**Comment by evlevy** [comment]
Can't remember... should this feature replace or augment min/maxDate in this package?

**Comment by lorejoh12** [comment]
> Can't remember... should this feature replace or augment min/maxDate in this package?

In the original spec discussions for it, we had considered implementations that would replace min/max date, but we didn't end up using them for simplicity's sake, we concluded the API was much more obvious with all 3 properties there.

File: packages/date-time/src/components/Calendar/CalendarDay/CalendarDay.base.tsx [hunk 2]
```diff
@@ -298,30 +298,79 @@ export class CalendarDayBase extends BaseComponent<ICalendarDayProps, ICalendarD
...
 
-    if (weekIndex === 0 && ev.which === KeyCodes.up) {
+    if (ev.which === KeyCodes.up) {
       targetDate = addWeeks(date, -1);
-    } else if (weekIndex === this.state.weeks!.length - 1 && ev.which === KeyCodes.down) {
+      direction = -1;
+    } else if (ev.which === KeyCodes.down) {
       targetDate = addWeeks(date, 1);
-    } else if (dayIndex === 0 && ev.which === getRTLSafeKeyCode(KeyCodes.left)) {
+      direction = 1;
```
Comment (evlevy) [inline packages/date-time/src/components/Calendar/CalendarDay/CalendarDay.base.tsx]: >direction = 1; [](start = 6, length = 14)

redundant

File: packages/date-time/src/components/Calendar/CalendarDay/CalendarDay.base.tsx [hunk 3]
```diff
@@ -298,30 +298,79 @@ export class CalendarDayBase extends BaseComponent<ICalendarDayProps, ICalendarD
...
+    } else if (ev.which === getRTLSafeKeyCode(KeyCodes.left)) {
       targetDate = addDays(date, -1);
-    } else if (dayIndex === DAYS_IN_WEEK - 1 && ev.which === getRTLSafeKeyCode(KeyCodes.right)) {
+      direction = -1;
+    } else if (ev.which === getRTLSafeKeyCode(KeyCodes.right)) {
       targetDate = addDays(date, 1);
+      direction = 1;
```
Comment (evlevy) [inline packages/date-time/src/components/Calendar/CalendarDay/CalendarDay.base.tsx]: >direction = 1; [](start = 6, length = 14)

redundant

File: packages/date-time/src/components/Calendar/examples/Calendar.Inline.Example.tsx (hunk 1 above)
Comment (lorejoh12) [reply packages/date-time/src/components/Calendar/examples/Calendar.Inline.Example.tsx]: this is actually causing an issue. If I change it to reference "@uifabric/date-time/lib/components/Calendar", the build fails with a "file does not exist". I'll add a work item for us to fix this later, but I'll revert the change for now.

Comment (lorejoh12) [reply packages/date-time/src/components/Calendar/examples/Calendar.Inline.Example.tsx]: opened #8146 for us to look in to

**Pull request merged**